cpp/src/arrow/csv/reader.cc

```diff
@@ -0,0 +1,399 @@
+
+class SerialTableReader : public BaseTableReader {
+ public:
+  SerialTableReader(MemoryPool* pool, std::shared_ptr<io::InputStream> input,
+                    const ReadOptions& read_options, const ParseOptions& parse_options,
+                    const ConvertOptions& convert_options)
+      : BaseTableReader(pool, read_options, parse_options, convert_options) {
+    // Since we're converting serially, no need to readahead more than one block
+    int32_t block_queue_size = 1;
+    readahead_ = std::make_shared<ReadaheadSpooler>(
+        pool_, input, read_options_.block_size, block_queue_size, kDefaultLeftPadding,
+        kDefaultRightPadding);
+  }
+
+  Status Read(std::shared_ptr<Table>* out) {
+    task_group_ = internal::TaskGroup::MakeSerial();
+
+    // First block
+    RETURN_NOT_OK(ReadNextBlock());
+    if (eof_) {
+      return Status::Invalid("Empty CSV file");
+    }
+    RETURN_NOT_OK(ProcessHeader());
+
+    auto parser =
+        std::make_shared<BlockParser>(parse_options_, num_cols_, read_options_.num_rows);
+    while (!eof_) {
+      // Consume current block
+      uint32_t parsed_size = 0;
+      RETURN_NOT_OK(parser->Parse(reinterpret_cast<const char*>(cur_data_),
+                                  static_cast<uint32_t>(cur_size_), &parsed_size));
+      if (parser->num_rows() > 0) {
+        // Got some data
+        RETURN_NOT_OK(ProcessData(parser, cur_block_index_++));
+        cur_data_ += parsed_size;
+        cur_size_ -= parsed_size;
+        if (!task_group_->ok()) {
+          // Conversion error => early exit
+          break;
+        }
+      } else {
+        // Need to fetch more data to get at least one row
+        RETURN_NOT_OK(ReadNextBlock());
+      }
+    }
+    if (eof_ && cur_size_ > 0) {
+      // Parse remaining data
+      uint32_t parsed_size = 0;
+      RETURN_NOT_OK(parser->ParseFinal(reinterpret_cast<const char*>(cur_data_),
+                                       static_cast<uint32_t>(cur_size_), &parsed_size));
+      if (parser->num_rows() > 0) {
+        RETURN_NOT_OK(ProcessData(parser, cur_block_index_++));
+      }
+    }
+
+    // Finish conversion, create schema and table
+    RETURN_NOT_OK(task_group_->Finish());
+    return MakeTable(out);
+  }
+};
+
+/////////////////////////////////////////////////////////////////////////
+// Parallel TableReader implementation
+
+class ThreadedTableReader : public BaseTableReader {
+ public:
+  ThreadedTableReader(MemoryPool* pool, std::shared_ptr<io::InputStream> input,
+                      ThreadPool* thread_pool, const ReadOptions& read_options,
+                      const ParseOptions& parse_options,
+                      const ConvertOptions& convert_options)
+      : BaseTableReader(pool, read_options, parse_options, convert_options),
+        thread_pool_(thread_pool) {
+    // Readahead one block per worker thread
+    int32_t block_queue_size = thread_pool->GetCapacity();
+    readahead_ = std::make_shared<ReadaheadSpooler>(
```

Comment:
Q: is ReadaheadSpooler zero copy on memory-mapped files?

Reply:
No, it isn't, because of padding.